ykrt/src/compile/mod.rs

```diff
@@ -28,6 +27,50 @@ pub(crate) mod jitc_llvm;
 #[cfg(jitc_yk)]
 pub mod jitc_yk;
 
+/// A failure to compile a trace.
+#[derive(Debug, thiserror::Error)]
+pub enum CompilationError {
+    #[error("Unrecoverable error: {0}")]
+    Unrecoverable(String),
+    #[error("Temporary error: {0}")]
+    Temporary(TemporaryErrorKind),
+}
+
+impl CompilationError {
+    pub fn new_jitir_index_overflow(which: String) -> Self {
+        Self::Temporary(TemporaryErrorKind::JITIRIndexOverflow(which))
+    }
+}
+
+pub type CompilationResult<T> = Result<T, CompilationError>;
```

Comment:
I must admit I really dislike `type` aliases used in this way: they save very little typing (in this case 6 characters) but now I have to know what they alias. So I find they make the code worse. I appreciate this is a personal quirk of mine so feel free to ignore if you disagree.

CompilationResult<T>
Result<T, CompilationError>

Reply:
I think this is one of the few cases where a type alias is acceptable. It's an idiom in use in the rust std lib, so I'm actually OK with it.

@ptersilie do you have any preference?

Reply:
Yes it is used in the standard library, but it really makes a mess there too IMHO.

Reply:
Changed in https://github.com/ykjit/yk/pull/956/commits/b169db4fbe9dcbb8e3b3bad5f4eeffd0e0b3ddd1